### PR TITLE
SHIP-0003: Tekton Custom Task

### PR DESCRIPTION
**Pull request by imjasonh**
cc @sbose78 @adambkaplan @qu1queee

### CODE REVIEW

File: ships/0003-tekton-custom-task.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+### Risks and Mitigations
+
+Integrating with Custom Tasks requires the Shipwright controller to have permission to view and update the status of _all `Run` resources_, even those that don't reference Shipwright `Build`s.
+This is due to limitations in Kubernetes RBAC -- there's no way to request access to only `Run` objects that reference `Build`s.
+As a result, a bug or exploit of the Shipwright controller could lead to unexpected or malicious behavior in the user's cluster.
+
+## Drawbacks
+
+This integration presents a somewhat convoluted semi-circular dependency between Shipwright and Tekton.
+A Tekton `PipelineRun` executes a `Run` that tells Shipwright to create a `BuildRun`, which executes by creating a Tekton `TaskRun` under the hood.
```
Comment (adambkaplan) [inline ships/0003-tekton-custom-task.md:103]: This is good to note for console implementations that want to visualize Shipwright builds. Consoles will always see a BuildRun mirror a TaskRun, so perhaps this isn't a huge issue?

**Review by adambkaplan — APPROVED** [review]
/approve

This is a good provisional proposal. My comments are meant for general discussion or items to consider when we get to the implementable phase.

File: ships/0003-tekton-custom-task.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+1. create a `Run` directly and check that a `BuildRun` is created; check that the `Run` status is updated when the `BuildRun` executes.
+1. create a `PipelineRun` that references such a `Run`, and check that the `PipelineRun` progresses and creates a `BuildRun`.
+
+### Release Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+#### Upgrade Strategy [if necessary]
+
+Currently, Tekton Custom Task `Run` resources have the `v1alpha1` API version.
```
[the remark this answers is not in the document]
Comment (imjasonh) [reply ships/0003-tekton-custom-task.md:92]: Good call, noted in the doc.

File: ships/0003-tekton-custom-task.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+## Motivation
+
+Implementing a Custom Task controller has two main benefits:
+
+1. This will allow us to integrate Shipwright with Tekton Triggers in a future change.
+   The Custom Task is necessary here because Tekton Triggers is only designed to create Tekton resources, and not arbitrary non-Tekton resources such as a `BuildRun`.
+1. This will allow users to integrate Shipwright Builds in their Tekton Pipelines.
+   This would enable them to, for example, run unit tests before calling Shipwright to build a container image, or to test, scan or deploy an image produced by Shipwright.
+
+Because Shipwright already necessarily requires Tekton to be installed, this change should not incur any additional dependencies on users or operators.
```
[the remark this answers is not in the document]
Comment (imjasonh) [reply ships/0003-tekton-custom-task.md:52]: v0.19.0 added initial support, so it shouldn't impose any more onerous Tekton version requirements than we already impose today. Noted in the doc.

File: ships/0003-tekton-custom-task.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+Currently, Tekton Custom Task `Run` resources have the `v1alpha1` API version.
+In the future, we should expect these to graduate to `v1beta1` and eventually `v1`.
+
+The shape of the `Run` resource is fairly small and not expected to grow in breaking ways.
+
+### Risks and Mitigations
+
+Integrating with Custom Tasks requires the Shipwright controller to have permission to view and update the status of _all `Run` resources_, even those that don't reference Shipwright `Build`s.
+This is due to limitations in Kubernetes RBAC -- there's no way to request access to only `Run` objects that reference `Build`s.
+As a result, a bug or exploit of the Shipwright controller could lead to unexpected or malicious behavior in the user's cluster.
```
[the remark this answers is not in the document]
Comment (imjasonh) [reply ships/0003-tekton-custom-task.md:101]: Yeah, it's unfortunately pretty unavoidable. I believe this is called out in Tekton's Custom Task EP as well.

File: ships/0003-tekton-custom-task.md (hunk 1 above)
Comment (imjasonh) [reply ships/0003-tekton-custom-task.md:106]: Good call, noted.

File: ships/0003-tekton-custom-task.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+Implement a Tekton Custom Task controller that watches for Tekton `Run` objects that reference a Shipwright `Build` resource by name.
+When such a `Run` object is created, create a `BuildRun` referencing the referenced `Build`, and as the `BuildRun` progresses, update the `Run` status accordingly.
+
+## Motivation
+
+Implementing a Custom Task controller has two main benefits:
+
+1. This will allow us to integrate Shipwright with Tekton Triggers in a future change.
+   The Custom Task is necessary here because Tekton Triggers is only designed to create Tekton resources, and not arbitrary non-Tekton resources such as a `BuildRun`.
```
Comment (sbose78) [inline ships/0003-tekton-custom-task.md:48]: +1

File: ships/0003-tekton-custom-task.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+1. This will allow users to integrate Shipwright Builds in their Tekton Pipelines.
+   This would enable them to, for example, run unit tests before calling Shipwright to build a container image, or to test, scan or deploy an image produced by Shipwright.
```
Comment (sbose78) [inline ships/0003-tekton-custom-task.md:50]: +1

File: ships/0003-tekton-custom-task.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbose78) [reply ships/0003-tekton-custom-task.md:101]: Yeah, good to have this noted down.

**Review by sbose78 — APPROVED** [review]
Looks good, would be nice to have a sample tekton pipeline yaml included in the proposal. :)

**Review by SaschaSchwarze0 — COMMENTED** [review]
I would like to make sure we make this feature meet the user's use cases. So, from a user's perspective, one might be missing certain features to integrate a Run which is a BuildRun into a larger Tekton pipeline:

* passing the source through a Tekton workspace; or at least overwriting the source (like the revision) of the build through the Pipeline
* getting insights into the image that was built (the image sha or the full image URL)

All these things require features in the build controller, but for the e2e story = to build a reasonable Tekton pipeline that does something useful, I think it will be necessary to speak out these things already.

**Comment by qu1queee** [comment]
@imjasonh sorry for the late reply in here. I would like this SHP to consider more aspects around the integration for:

- Tekton Triggers
- An embedded SHP in Tekton Pipeline resources

I do not have the whole picture yet clear, but I wrote some key points we might wanna consider, see [delta](https://github.com/imjasonh/community-2/compare/tekton-custom-task...qu1queee:qu1queee/shp003-feedback?expand=1&short_path=8e9ca89#diff-8e9ca89d5011659400100d5a71cf5e4cd22c03dc80353820f599d53b21792adf). I can open a proper PR to yours if you want, or you can consume some of the ideas I wrote there. Let me know.

**Review by qu1queee — CHANGES_REQUESTED** [review]
@imjasonh sorry for the late reply in here. I would like this SHP to consider more aspects around the integration for:

- Tekton Triggers
- An embedded SHP in Tekton Pipeline resources

I do not have the whole picture yet clear, but I wrote some key points we might wanna consider, see [delta](https://github.com/imjasonh/community-2/compare/tekton-custom-task...qu1queee:qu1queee/shp003-feedback?expand=1&short_path=8e9ca89#diff-8e9ca89d5011659400100d5a71cf5e4cd22c03dc80353820f599d53b21792adf). I can open a proper PR to yours if you want, or you can consume some of the ideas I wrote there. Let me know.

**Comment by imjasonh** [comment]
Hi @qu1queee @SaschaSchwarze0 sorry for the _very_ late reply. 😓 

I agree there's probably more design work to be done to think through how this integrates with a Tekton Pipeline, especially the areas you've highlighted around reusing Tekton's `workspaces`, and emitting `results` to downstream Tasks. I believe with changes like #13 we can have the Custom Task controller adapt Shipwright Build `.status.results` (incl git SHA and image SHA) to Tekton `.status.results`, and with the evolving work around `sources` we can find some way to adapt `workspaces` to an input `source`.

Ultimately this is a `provisional` SHIP, so I don't think we need to answer all of these questions definitively before merging it. We can layer on more design considerations and features before it becomes `implementable`.

**Comment by adambkaplan** [comment]
@imjasonh since we merged the previous enhancement proposals here, please update the title and numbering to SHIP-0020

**Comment by qu1queee** [comment]
@imjasonh moving out of this PR as I would not be able to provide feedback in the next weeks. @SaschaSchwarze0 fyi.

**Pull request closed without merging**